File: Javascript/Two Pointers/M-11. Container With Most Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  const n = height.length;
  let max = -Number.MAX_SAFE_INTEGER;
  let l = 0;
  let r = n - 1;
  while (l < r) {
    const current = (r - l) * Math.min(height[l], height[r]);
    max = Math.max(max, current);
    if (height[l] < height[r]) {
      l++;
    } else {
      r--;
    }
  }
  return max;
};

const testData = [
  {
    height: [1, 8, 6, 2, 5, 4, 8, 3, 7],
    ans: 49,
  },
  {
    height: [1, 1],
    ans: 1,
  },
];

for (const data of testData) {
  console.log(data.ans === maxArea(data.height));
}

File: Python/Two Pointers/M-11. Container With Most Water.py
import sys


class Solution:
    @classmethod
    def maxArea(self, height) -> int:
        max_area = -sys.maxsize
        n = len(height)
        l = 0
        r = n - 1

        while l < r:
            cur_area = (r - l) * min(height[l], height[r])
            max_area = max(max_area, cur_area)
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1

        return max_area


test_data = [
    {
        "height": [1, 8, 6, 2, 5, 4, 8, 3, 7],
        "ans": 49,
    },
    {
        "height": [1, 1],
        "ans": 1,
    },
]

for data in test_data:
    print(data["ans"] == Solution.maxArea(data["height"]))
